### PR TITLE
`sz analyze` command refactoring.

### PR DESCRIPTION
**Pull request by Jonas-Sander**
I refactored the code for the `sz analyze` command so that the logic for concurrently running a task for all packages can be used for the test command as well (#340).

### CODE REVIEW

File: tools/sz_repo_cli/lib/src/common/src/concurrent_package_task_runner.dart [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    ///
+    /// Most often this will be [SharezoneRepo.streamPackages], maybe with a
+    /// filter (e.g. `.where((package) => package.hasTestDirectory`) or
+    /// something similar.
+    @required Stream<Package> packageStream,
+    @required Future<void> Function(Package package) runTask,
+
+    /// If 0, null or negative then every package will be processed at once.
+    /// Else [maxNumberOfPackagesBeingProcessedConcurrently] are processed
+    /// concurrently,
```
Comment (nilsreichardt) [inline tools/sz_repo_cli/lib/src/common/src/concurrent_package_task_runner.dart:39]:
```suggestion
    /// Defines the number of packages that being processed concurrently.
    ///
    /// If 0, null or negative then every package will be processed at once.
```

Even when the first sentence repeats the name of the variable, is this maybe a better documentation that follows Effective Dart in the following points:
* [DO start doc comments with a single-sentence summary.](https://dart.dev/guides/language/effective-dart/documentation#do-start-doc-comments-with-a-single-sentence-summary)
* [DO separate the first sentence of a doc comment into its own paragraph.](https://dart.dev/guides/language/effective-dart/documentation#do-separate-the-first-sentence-of-a-doc-comment-into-its-own-paragraph)

File: tools/sz_repo_cli/lib/src/common/src/concurrent_package_task_runner.dart [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+    ));
+    _controller.close();
+  }
+}
+
+class PackageTask {
+  final Package package;
+  final Stream<PackageTaskStatus> status;
+
+  PackageTask(this.package, this.status);
```
Comment (nilsreichardt) [inline tools/sz_repo_cli/lib/src/common/src/concurrent_package_task_runner.dart:181]:
```suggestion
  const PackageTask(this.package, this.status);
```

File: tools/sz_repo_cli/lib/src/common/src/concurrent_package_task_runner.dart [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+}
+
+abstract class PackageTaskStatus {
+  final Package package;
+  final DateTime startedOn;
+
+  PackageTaskStatus({
```
Comment (nilsreichardt) [inline tools/sz_repo_cli/lib/src/common/src/concurrent_package_task_runner.dart:188]:
```suggestion
  const PackageTaskStatus({
```

File: tools/sz_repo_cli/lib/src/common/src/concurrent_package_task_runner.dart [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+      .toList();
+
+  Future<List<Success>> get allSuccesses => map((event) => event.status)
+      .asyncExpand((status) => status)
+      .whereType<Success>()
+      .toList();
+}
+
+class PackageTasksStatusPresenter {
+  PackageTasksStatusPresenter();
```
Comment (nilsreichardt) [inline tools/sz_repo_cli/lib/src/common/src/concurrent_package_task_runner.dart:281]:
```suggestion
  const PackageTasksStatusPresenter();
```

File: tools/sz_repo_cli/pubspec.yaml [hunk 5]
```diff
@@ -10,26 +10,26 @@ name: sz_repo_cli
...
   union: ^0.0.3+1
-  optional: ^4.1.0
-  console: ^3.1.0
-  rxdart: ^0.27.1
-
-environment:
-  sdk: ">=2.10.0 <3.0.0"
+  optional: ^6.1.0+1
+  console: ^4.1.0
+  rxdart: ^0.27.1
```
Comment (nilsreichardt) [inline tools/sz_repo_cli/pubspec.yaml:35]:
```suggestion
  rxdart: ^0.27.1

```

Nit: At newline at end of file

File: tools/sz_repo_cli/lib/src/common/src/sharezone_repo.dart [hunk 6]
```diff
@@ -50,7 +50,45 @@ class SharezoneRepo {
   }
 
   Stream<Package> streamPackages() {
-    return dartLibraries.streamPackages().mergeWithValues([
+    // This stream is often used to start tasks for all packages (like testing
```
Comment (nilsreichardt) [inline tools/sz_repo_cli/lib/src/common/src/sharezone_repo.dart:53]: Thanks for the comment ❤️

**Review by nilsreichardt — APPROVED** [review]
LGTM 👍

File: tools/sz_repo_cli/lib/src/common/src/concurrent_package_task_runner.dart [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+class ConcurrentPackageTaskRunner {
+  DateTime Function() _getCurrentDateTime;
+
+  ConcurrentPackageTaskRunner({
+    DateTime Function() getCurrentDateTime,
+  }) {
+    _getCurrentDateTime = getCurrentDateTime ?? () => DateTime.now();
+  }
+
+  /// Run [runTask] for every package from the [packageStream].
```
Comment (Jonas-Sander) [inline tools/sz_repo_cli/lib/src/common/src/concurrent_package_task_runner.dart:27]:
```suggestion
  /// Run [runTask] for every [Package] emitted by the [packageStream].
```

File: tools/sz_repo_cli/lib/src/common/src/concurrent_package_task_runner.dart (hunk 1 above)
Comment (Jonas-Sander) [reply tools/sz_repo_cli/lib/src/common/src/concurrent_package_task_runner.dart:39]: Thanks!

**Pull request merged**